packages/meditrak-server/src/routes/GETMapOverlays/assertMapOverlaysPermissions.js

```diff
@@ -2,72 +2,64 @@
-    const mapOverlayEntities = countryCodes.map(c => entityByCode[c]);
-    const permissionGroup = mapOverlay.userGroup;
-    const countryCodesString = countryCodes.join(',');
-
-    //permissionGroup and countryCodes are the 2 values for checking access for a map overlay
-    const cacheKey = `${permissionGroup}/${countryCodesString}`;
-    let hasAccessToMapOverlay = accessCache[cacheKey];
-
-    //Perform access checking if its not in the cache, otherwise reuse the value
-    if (hasAccessToMapOverlay === undefined) {
-      for (let j = 0; j < mapOverlayEntities.length; j++) {
-        const mapOverlayEntity = mapOverlayEntities[j];
-        hasAccessToMapOverlay = await hasAccessToEntityForVisualisation(
-          accessPolicy,
-          models,
-          mapOverlayEntity,
-          permissionGroup,
-        );
+const { RAW } = QUERY_CONJUNCTIONS;
 
-        //If users have access to any countries of the overlay,
-        //it should be enough to allow access the overlay
-        if (hasAccessToMapOverlay) {
-          break;
-        }
-      }
+export const assertMapOverlaysPermissions = async (accessPolicy, models, mapOverlayId) => {
+  const mapOverlay = await models.mapOverlay.findById(mapOverlayId);
 
-      accessCache[cacheKey] = hasAccessToMapOverlay;
-    }
+  const countryCodes = mapOverlay.countryCodes || [];
+  const entities = await models.entity.find({ code: countryCodes });
 
-    if (hasAccessToMapOverlay) {
-      filteredMapOverlays.push(mapOverlay);
+  for (let i = 0; i < entities.length; i++) {
+    if (
+      await hasAccessToEntityForVisualisation(
+        accessPolicy,
+        models,
+        entities[i],
+        mapOverlay.userGroup,
+      )
+    ) {
+      return true;
     }
   }
 
-  return filteredMapOverlays;
+  throw new Error('You do not have permissions for the requested map overlay(s)');
 };
 
-export const assertMapOverlaysPermissions = async (accessPolicy, models, mapOverlays) => {
-  const filteredMapOverlays = await filterMapOverlaysByPermissions(
-    accessPolicy,
-    models,
-    mapOverlays,
-  );
-
-  if (filteredMapOverlays.length !== mapOverlays.length) {
-    throw new Error('You do not have permissions for the requested map overlay(s)');
+export const createMapOverlayDBFilter = async (accessPolicy, models, criteria) => {
+  if (hasBESAdminAccess(accessPolicy)) {
+    return criteria;
   }
+  const dbConditions = criteria;
+  const allPermissionGroups = accessPolicy.getPermissionGroups();
+  const countryCodesByPermissionGroup = {};
+
+  // Generate lists of country codes we have access to per permission group
+  allPermissionGroups.forEach(pg => {
+    countryCodesByPermissionGroup[pg] = accessPolicy.getEntitiesAllowed(pg);
+  });
 
-  return true;
+  // Look up the json country codes by userGroup, compare them to the countryCodes from the mapOverlay
+  dbConditions[RAW] = {
+    sql: `("countryCodes" && array(select trim('"' from json_array_elements(?::json->"userGroup")::text)))
```

Comment:
SQL is impressive once again! Could you add a bit of whitespace to ease the reading experience?